### PR TITLE
chore: remove hvd_config usage [DET-6220]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

Separating this change from native pytorch dtrain support.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: harness/determined/estimator/_estimator_context.py
Comment (rb-determined-ai) [inline harness/determined/estimator/_estimator_context.py:47]: EstimatorTrial can't run with non-horovod distributed backend.  If you ran it with a different dtrain backend, what you have here would silently exhibit weird behavior.  The check you have here follows more closely to the pre-existing code paths, which I assume is why you implemented it this way.  But the pre-existing code paths implicitly assumed that we only ever ran horovod train.  Now that the system no longer enforces that restriction globally, what you have here doesn't make sense in isolation.

As it is right now, if I read this in isolation, I would say, "why does average_training_metrics, something that is common to all dtrain, not apply with to a tensorflow-native distributed backend?  Why only horovod?"

I of course don't want you to add non-horovod support to EstimatorTrial.  I would recommend writing sections like this as `if self.context.distributed.size > 1` (which is equivalent, since only horovod is supported) and I would simply add a check in harness.py that throws an error if somebody tries to launch EstimatorTrial or TFKerasTrial without a horovod launch layer.  With this strategy, there's no need to pass `distributed_backend` into EstimatorTrial or TFKerasTrial at all.

File: harness/determined/exec/harness.py
Comment (rb-determined-ai) [inline harness/determined/exec/harness.py:81]: This looks like clear evidence that you could use `if context.distributed.size > 1` everywhere that `if hvd_config.use` is used

File: harness/determined/pytorch/_pytorch_context.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_context.py:89]: Do you mean for these to be user-facing APIs?  If not they need to each have an underscore prefix.

There are some old attributes that didn't get this treatment but also aren't really public APIs, like context.env and context.experiment_config, but we should be consistent with new values.

**Review by rb-determined-ai — COMMENTED** [review]
Ultimately, this pattern of having a `distributed_backend` which is a string that everything in our code knows how to interpret is the reverse of what we want to build.

This pattern means that all of the objects in our code need to know how to deal with each value that `distributed_backend` could take.  That's a non-extensible interface, because users might be able to set a different string but they couldn't teach all of the objects in our code how to deal with it.

A better pattern is to write an interface that each backend could provide, and let all of the objects in our code accept an object that implements that interface.  That way, if a user wants to extend the system, they could implement that interface themselves, and our system would continue to operate normally.

In this case, you don't need to introduce a new interface type (please don't, in fact); I'm pretty sure you can get away with just passing the `DistributedContext` around, since that provides most of the features that we need, at least for side-features like the data layer.

This leaves one important exception though: how will you deal with PyTorchTrial, where you do need to know backend information?  Any ideas?

**Comment by azhou-determined** [comment]
> A better pattern is to write an interface that each backend could provide, and let all of the objects in our code accept an object that implements that interface. That way, if a user wants to extend the system, they could implement that interface themselves, and our system would continue to operate normally.
> 

> In this case, you don't need to introduce a new interface type (please don't, in fact); I'm pretty sure you can get away with just passing the `DistributedContext` around, since that provides most of the features that we need, at least for side-features like the data layer.
> 

agreed. after some experimentation of different implementations for an interface, I've settled on the environment variable approach. 

each launch layer will set an environment variable flag (USE_HOROVOD = True/False and subsequently USE_PYTORCH_DTRAIN) indicating which distributed backend is being used, which will be read/interpreted by the harness and trial classes. 

this is nice because it logically separates out the code concerns of distributed backend and multi-GPU training: i.e. we could be running a horovod backend but only using one GPU (USE_HOROVOD = True and context.distributed.size = 1)

for features like data layer that only need to know whether multi-GPU training is being used, passing in the DistributedContext is enough. for class methods like pre_execute_hook, it only needs to know which backend(s) is being run.

this will also allow "layering" of launch layers in the future, like if you wanted to run torch distributed along with a custom profiling layer or etc, while keeping if/else statements clean-ish and to a minimum.

while setting too many environment variables can get messy, i think this is the best approach moving forward. other implementations get tangled up with various trial and distributed contexts and inevitably get glued together with scattered if/else logic.

File: harness/determined/_trial_controller.py
Comment (azhou-determined) [inline harness/determined/_trial_controller.py:13]: this needs a better place to live. I don't feel like it belongs in trial context or harness though

File: harness/determined/keras/_tf_keras_trial.py
Comment (rb-determined-ai) [inline harness/determined/keras/_tf_keras_trial.py:215]: just use `self.context.distributed.local_rank` here, it'll be easier.

File: harness/determined/keras/_tf_keras_trial.py
Comment (rb-determined-ai) [inline harness/determined/keras/_tf_keras_trial.py:509]: as with estimator trial, I'd prefer you put an assert somewhere that only allows dtrain tfkeras when horovod is in use.

If we ever support non-horovod dtrain tfkeras then we'll already have to go through the code with a fine-tooth comb, and these if statements where we try to predict what the right behavior would be if we supported non-horovod dtrain tfkeras would be more likely to cause quiet bugs than to be helpful.

File: harness/determined/keras/_tf_keras_trial.py
Comment (rb-determined-ai) [inline harness/determined/keras/_tf_keras_trial.py:822]: ditto

File: harness/determined/keras/_tf_keras_trial.py
Comment (rb-determined-ai) [inline harness/determined/keras/_tf_keras_trial.py:879]: ditto

File: harness/determined/keras/_tf_keras_trial.py
Comment (rb-determined-ai) [inline harness/determined/keras/_tf_keras_trial.py:916]: ditto

File: harness/determined/pytorch/_pytorch_context.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_context.py:211]: this line seems pointless.

File: harness/determined/pytorch/_pytorch_trial.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:444]: just use `self.context.distributed.size` (you have quite a few of these)

File: harness/determined/pytorch/_pytorch_trial.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:439]: Same thing about trying to distinguish horovod dtrain from non-horovod dtrain.  As evidence of my claim that "trying to predict the horovod codepaths and the non-horovod codepaths before you actually add a second dtrain backend will mostly cause bugs", you have not been consistent in these checks even in this PR.  For example, context.wrap_optimizer uses horovod but does not have protection.  Having some checks now is more of a false sense of security, IMO.

However, since you are actively working on non-horovod dtrain for pytorch trial, it's really up to you if you want to get rid of these horovod-specific checks with pytorch trial.  You do have enough background to make correct decisions here, which will be meaningful in the near future.  You can take my suggestions here or not.  With TFKeras or Estimator though, I would definitely like you to remove the horovod-specific checks and just have assertions that only the horovod backend is in use.  I really doubt that we ever need to revisit those trial classes, but if we do, it would be better to start with a blank slate, so to speak.

File: harness/determined/pytorch/_pytorch_trial.py
Comment (rb-determined-ai) [inline harness/determined/pytorch/_pytorch_trial.py:271]: same as before on replacing `check.true` with an assertion.  If you are touching this line, please make this update as well.

**Review by rb-determined-ai — COMMENTED** [review]
Overall, looks pretty good.  Lots of pretty small feedback items, with only one incorrect thing that I could find.

File: harness/determined/pytorch/_pytorch_trial.py
Comment (azhou-determined) [reply harness/determined/pytorch/_pytorch_trial.py:439]: since I want to get this change out before introducing native dtrain for pytorch, I went with your suggestion of removing horovod checks and asserting horovod must be used if distributed training.

File: harness/determined/keras/_tf_keras_trial.py
Comment (azhou-determined) [reply harness/determined/keras/_tf_keras_trial.py:215]: this method is a @classmethod, can't use self... methods for class instances.

File: harness/determined/_trial_controller.py
Comment (azhou-determined) [reply harness/determined/_trial_controller.py:5]: why?

File: harness/determined/estimator/_estimator_trial.py
Comment (azhou-determined) [reply harness/determined/estimator/_estimator_trial.py]: can't access self instance variables on a @classmethod

File: harness/determined/_trial_controller.py
Comment (rb-determined-ai) [reply harness/determined/_trial_controller.py:4]: please read the google style guide for imports, as linked before.

File: harness/determined/_data_layer/_context.py
Comment (rb-determined-ai) [inline harness/determined/_data_layer/_context.py:4]: from determined import _generic

Prefer to import the final module as a single name.

File: harness/determined/_data_layer/_data_layer.py
Comment (rb-determined-ai) [inline harness/determined/_data_layer/_data_layer.py:13]: from determined import _generic
...
_generic.DistributedContext

Prefer to import the final module as a single name (and only import modules, not symbols)

File: harness/determined/_trial_controller.py
Comment (rb-determined-ai) [inline harness/determined/_trial_controller.py:12]: Please see the previous comment about this being an enum.  I marked it as unresolved.

Also, please don't mark something as resolved unless you are 100% sure you've fixed it.  IMO it's better not to mark something as resolved until the review has had a chance to see how you've fixed it, then the reviewer can mark it as resolved.  But not everybody agrees with that.

But definitely don't mark something as resolved when it's not fixed.

File: harness/determined/estimator/_estimator_context.py
Comment (rb-determined-ai) [inline harness/determined/estimator/_estimator_context.py:9]: from determined import _generic
...
_generic.DistributedContext

File: harness/determined/estimator/_estimator_trial.py
Comment (rb-determined-ai) [inline harness/determined/estimator/_estimator_trial.py:412]: I can't recall anymore what your last version had here.  It seems like probably you want to use the new _DistributedBackend here maybe?  Seems to carry the information you're trying to portray here?

I'd consider this a "non-blocking" review comment.  Ignore it if you wish.

File: harness/determined/keras/_tf_keras_context.py
Comment (rb-determined-ai) [inline harness/determined/keras/_tf_keras_context.py:9]: from determined import _generic
...
_generic.DistributedContext

**Review by rb-determined-ai — APPROVED** [review]
Very close now.

I'll approve this since I'm on vacation, but please do make the touchups I suggested.

Also before you land this, since you are touching so much of the multi-gpu codepath you need to run the upstream tests first:
* create an upstream PR (push to master/deprecate-hvd-config).  Make sure to alter the commit hash before pushing (git commit --amend) because sometimes I've seen circleci be weird if there are two prs with identical commit hashes trying to run tests
* Enable all of the multi-gpu tests manually.  Don't bother with the nightly or convergence tests or weekly tests or whatever they're called.  Or the model hub tests.  Those are unreasonably slow for this.
* Wait for those to pass on the upstream PR, then land this PR (with identical code, of course).  That is better for the git history, since this PR will have all of the review comments.
* If you have any questions about this process, ask @ioga.

**Comment by azhou-determined** [comment]
> 

Made suggested changes and deploying upstream tests. GKE is failing for unrelated reasons, but other GPU tests seem to be running fine. Going to monitor/rerun a few times, and if all looks well, I will merge.

**Comment by azhou-determined** [comment]
> Very close now.
> 
> I'll approve this since I'm on vacation, but please do make the touchups I suggested.
> 
> Also before you land this, since you are touching so much of the multi-gpu codepath you need to run the upstream tests first:
> 
> * create an upstream PR (push to master/deprecate-hvd-config).  Make sure to alter the commit hash before pushing (git commit --amend) because sometimes I've seen circleci be weird if there are two prs with identical commit hashes trying to run tests
> * Enable all of the multi-gpu tests manually.  Don't bother with the nightly or convergence tests or weekly tests or whatever they're called.  Or the model hub tests.  Those are unreasonably slow for this.
> * Wait for those to pass on the upstream PR, then land this PR (with identical code, of course).  That is better for the git history, since this PR will have all of the review comments.
> * If you have any questions about this process, ask @ioga.

upstream tests are passing here (https://app.circleci.com/pipelines/github/determined-ai/determined/17468/workflows/e635e4d4-a4b9-491a-8513-c8160cc79a7c). some intermittent failures on GKE clusters spinning up, but seems like no errors related to this change. going to land this.

**Pull request merged**
